Refactor BasecallsConverter

jacarey opened the pull request.

Commit:
Normalize data provider entry points
Unified basecalls converter

Commit:
First pass at BasecallsConverterBuilder

Commit:
Pass two for BasecallsConverterBuilder

Commit:
Fix bug in tile fetch for multi tile data.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -1,21 +1,36 @@
+    public static final IlluminaDataType[] DATA_TYPES_WITHOUT_BARCODE =
+            Arrays.copyOfRange(DATA_TYPES_WITH_BARCODE, 0, DATA_TYPES_WITH_BARCODE.length - 1);
 
+    protected final BclQualityEvaluationStrategy bclQualityEvaluationStrategy;
```
Given that this class now has no subclasses, can we try and make as many of the members private as possible please?

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -26,38 +41,54 @@
     final ProgressLogger writeProgressLogger = new ProgressLogger(log, 1000000, "Write");
+    final boolean includeNonPfReads;
+    private final Map<String, ThreadPoolExecutorWithExceptions> barcodeWriterThreads = new HashMap<>();
+    private final Map<Integer, List<RecordWriter>> completedWork = new HashMap<>();
```
I'm assuming this and the next are `Map<Integer,?>` because they are keyed by tile, and tiles are not sequential starting at 0?  That said, if tile numbers don't get hilariously large it would likely be much more efficient at the cost of a little memory to store these as arrays that just have lots of gaps in them.

jacarey replied:
MiSeq tile numbers get into the 10s of thousands. And there are never all that many tiles total. So I think a Map is likely a decent tradeoff for memory vs efficiency. I think we could likely determine the max tile name and instantiate an array of that size and it shouldn't get too bad, but if Illumina changes tile number it could spiral

tfenne replied:
Ah well ok then.  And thinking more I think the maps of tile -> things should be accessed a lot less frequently than maps of barcode -> things, so let's just ignore the tile maps then.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -26,38 +41,54 @@
+    private final Map<String, ThreadPoolExecutorWithExceptions> barcodeWriterThreads = new HashMap<>();
+    private final Map<Integer, List<RecordWriter>> completedWork = new HashMap<>();
+    private final Map<Integer, File> barcodesFiles = new HashMap<>();
+    protected List<Integer> tiles;
```
Ditto here that this could probably be an `int[]` instead?

jacarey replied:
I went down this rabbit hole with no appreciable efficiency change.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -26,38 +41,54 @@
      * @param tmpDirs                  For SortingCollection spilling.
      * @param numProcessors            Controls number of threads.  If <= 0, the number of threads allocated is
      *                                 available cores - numProcessors.
+     * @param firstTile                (For infoging) If non-null, start processing at this tile.
```
Lol .. I think someone (you? someone else?) must have done a global search/replace for debug=>info!

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -26,38 +41,54 @@
-                       final int numProcessors,
-                       final IlluminaDataProviderFactory factory) {
-
+    public BasecallsConverter(final File basecallsDir, final File barcodesDir, final int lane,
```
Can we put the first three params on their own lines too please?

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -77,14 +108,52 @@
+        }
+    }
+
+    public static File[] getTiledFiles(final File baseDirectory, final Pattern pattern) {
```
Could this be private?  Or at least package access?

jacarey replied:
It is used by `IlluminaDataProviderFactory` which is in picard.illumina.parser package. We could reorg.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -77,14 +108,52 @@
+        this.tiles = factory.getAvailableTiles();
+        tiles.sort(TILE_NUMBER_COMPARATOR);
+        setTileLimits(firstTile, tileLimit);
+        barcodeRecordWriterMap.keySet().forEach(barcode -> barcodeWriterThreads.put(barcode, new ThreadPoolExecutorWithExceptions(1)));
```
Yikes - so we're creating a thread per barcode?  Is this also true in the current implementation or is this something you've added?

jacarey replied:
The new converter has always been that way. The old provider uses a prioritizing thread pool. We could use either one.
I definitely did this before, not this time around, but this is likely causing a lot of thrashiing.

I'm going to switch this over to using a shared thread pool executor.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -1,21 +1,36 @@
+import java.util.regex.Pattern;
 
-abstract class BasecallsConverter<CLUSTER_OUTPUT_RECORD> {
+public class BasecallsConverter<CLUSTER_OUTPUT_RECORD> {
```
Can you please add a decent class level javadoc?  I think if we spend a little time going through and clearly documenting classes and methods we touch in this PR, it's a nice show of good faith and will likely help get the PR accepted in the main repo.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -113,6 +182,50 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
         }
     }
 
+    public void doTileProcessing() {
```
Javadoc please.  Also I wonder if this method's name should change to something a bit more descriptive?  It's basically `doAllTheThings()`, so perhaps something like `processTilesAndWritePerSampleOutputs()`?

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -113,6 +182,50 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+
+        final CompletedWorkChecker workChecker = new CompletedWorkChecker();
+        completedWorkExecutor.submit(workChecker);
+        completedWorkExecutor.shutdown();
```
```suggestion
        final ThreadPoolExecutorWithExceptions completedWorkExecutor = new ThreadPoolExecutorWithExceptions(1);
        final CompletedWorkChecker workChecker = new CompletedWorkChecker();
        completedWorkExecutor.submit(workChecker);
        completedWorkExecutor.shutdown();
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -113,6 +182,50 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+            tileProcessingExecutor.submit(new TileProcessor(tile, barcodesFiles.get(tile)));
+        }
+
+        tileProcessingExecutor.shutdown();
```
```suggestion
        final ThreadPoolExecutorWithExceptions tileProcessingExecutor = new ThreadPoolExecutorWithExceptions(numThreads);
        for (final Integer tile : tiles) {
            tileProcessingExecutor.submit(new TileProcessor(tile, barcodesFiles.get(tile)));
        }
        tileProcessingExecutor.shutdown();
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
-    /**
-     * A comparator for tile numbers, which are not necessarily ordered by the number's value.
-     */
+    private class RecordWriter implements Runnable {
```
Javadoc please!

tfenne replied:
Also, a better name would be nice.  While this does write records, with all the other classes flying around (e.g. ClusterDataWriter) it would be nice to clarify that this is a `Runnable` that "drains" a sorting collection to the output writer.  How about `SortedRecordToWriterPump` or something along those lines?

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+        private final String barcode;
+
+        RecordWriter(final ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD> writer,
+                     final SortingCollection<CLUSTER_OUTPUT_RECORD> recordCollection, final String barcode) {
```
```suggestion
        RecordWriter(final ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD> writer,
                     final SortingCollection<CLUSTER_OUTPUT_RECORD> recordCollection, 
                     final String barcode) {
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+            log.debug("Closing writer for barcode " + barcode);
+            this.writer.close();
+        }
+    }
```
I miss scala where this would be ~2-3 lines of code.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+        }
+    }
+
+    private class TileProcessor implements Runnable {
```
Javadoc please.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+
+    private class TileProcessor implements Runnable {
+        private final int tileNum;
+        private final Map<String, SortingCollection<CLUSTER_OUTPUT_RECORD>> barcodeToRecordCollection = new HashMap<>();
```
Not so much for this PR, but thinking aloud, I wonder if it is worth spending a little time finding the most efficient possible `Map` for ~8-5000 fixed-length string keys for places like this?  Presumably we're hitting this map and maps like this at least once per cluster and likely many times.

tfenne replied:
Thinking ahead, I think we should introduce a generic writer/sink interface in `HTSJDK` that is similar to this:
    https://github.com/fulcrumgenomics/commons/blob/master/src/main/scala/com/fulcrumgenomics/commons/io/Writer.scala

with methods to write one record, an iterator of records and an iterable of records, and extend closeable/autocloseable.  We could then make `SortingCollection` extend that, and all the writer classes in BasecallsConverter.  Then in places like this we could take a map where the values extend `Writer<CLUSTER_OUTPUT_RECORD>` instead of `SortingCollection`. 

Doesn't necessarily have to be in this PR, but maybe we should start that in parallel?

jacarey replied:
We know the map size ahead of time since we know our total number of barcode writers we need. I suspect the most efficient map is going to be `new Hashmap<>(numBarcodes, 1.0f)`

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+
+        @Override
+        public void run() {
+            final BaseIlluminaDataProvider dataProvider = factory.makeDataProvider(tileNum, barcodeFile);
```
Do you think there is any efficiency to be gained by pushing the `pf only` down into the data provider?  I.e. would they be able to skip over non-pf reads without doing all the decoding work?  Maybe add that to a list of possible future things to try?

jacarey replied:
Lets add that to the list. For newer machines this won't have much of an impact, but it could be substantial for others.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+                final String barcode = (demultiplex ? cluster.getMatchedBarcode() : null);
+                if (cluster.isPf() || includeNonPfReads) {
+                    addRecord(barcode, converter.convertClusterToOutputRecord(cluster));
+                }
```
```suggestion
                if (cluster.isPf() || includeNonPfReads) {
                    final String barcode = (demultiplex ? cluster.getMatchedBarcode() : null);
                    addRecord(barcode, converter.convertClusterToOutputRecord(cluster));
                }
```

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+
+        private synchronized void addRecord(final String barcode, final CLUSTER_OUTPUT_RECORD record) {
+            // Grab the existing collection, or initialize it if it doesn't yet exist
+            SortingCollection<CLUSTER_OUTPUT_RECORD> recordCollection = this.barcodeToRecordCollection.get(barcode);
```
Could we just initialize all these in the constructor and make life simpler here?  I guess it's possible that some tiles may not have reads for some barcodes so we may end up making more than we need, but that seems like an edge case.  It would make this simpler and also make the check for `ignoreUnexpectedBarcodes` better too.

tfenne commented on src/main/java/picard/illumina/BasecallsConverter.java:
```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
+    }
+
+
+    private class CompletedWorkChecker implements Runnable {
```
Javadoc please.

tfenne commented on src/main/java/picard/illumina/BasecallsConverterBuilder.java:
```diff
@@ -0,0 +1,119 @@
+import java.util.List;
+import java.util.Map;
+
+public class BasecallsConverterBuilder<CLUSTER_OUTPUT_RECORD> {
```
Javadoc please.

tfenne replied:
On all the public methods too please.  Sorry.

tfenne commented on src/main/java/picard/illumina/BasecallsConverterBuilder.java:
```diff
@@ -0,0 +1,119 @@
+
+public class BasecallsConverterBuilder<CLUSTER_OUTPUT_RECORD> {
+
+    File basecallsDir;
+    File barcodesDir;
+    int lane;
+    ReadStructure readStructure;
+
+    Comparator<CLUSTER_OUTPUT_RECORD> outputRecordComparator;
+    SortingCollection.Codec<CLUSTER_OUTPUT_RECORD> codecPrototype;
+    Class<CLUSTER_OUTPUT_RECORD> outputRecordClass;
+
+    Map<String, ? extends BasecallsConverter.ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD>> barcodeRecordWriterMap;
+    boolean demultiplex = false;
+    int maxReadsInRamPerTile = 1200000;
+    List<File> tmpDirs = Collections.singletonList(new File(System.getProperty("java.io.tmpdir")));
+    int numProcessors = Runtime.getRuntime().availableProcessors();
+    Integer firstTile = null;
+    Integer tileLimit = null;
+    BclQualityEvaluationStrategy bclQualityEvaluationStrategy =
+            new BclQualityEvaluationStrategy(BclQualityEvaluationStrategy.ILLUMINA_ALLEGED_MINIMUM_QUALITY);;
+    boolean ignoreUnexpectedBarcodes = false;
+    boolean applyEamssFiltering = false;
+    boolean includeNonPfReads = false;
```
Can these all be private?

tfenne commented on src/main/java/picard/illumina/BasecallsConverterBuilder.java:
```diff
@@ -0,0 +1,119 @@
+    Map<String, ? extends BasecallsConverter.ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD>> barcodeRecordWriterMap;
+    boolean demultiplex = false;
+    int maxReadsInRamPerTile = 1200000;
+    List<File> tmpDirs = Collections.singletonList(new File(System.getProperty("java.io.tmpdir")));
```
Is there not a more picard-way to get a tmp dir?  Something that would respect `TMP_DIR=` from the command line?

tfenne commented on src/main/java/picard/illumina/BasecallsConverterBuilder.java:
```diff
@@ -0,0 +1,119 @@
+    Integer firstTile = null;
+    Integer tileLimit = null;
+    BclQualityEvaluationStrategy bclQualityEvaluationStrategy =
+            new BclQualityEvaluationStrategy(BclQualityEvaluationStrategy.ILLUMINA_ALLEGED_MINIMUM_QUALITY);;
```
a) love the "alleged", b) double semi-colon

tfenne commented on src/main/java/picard/illumina/CheckIlluminaDirectory.java:
```diff
@@ -116,7 +116,7 @@
         final ReadStructure readStructure = new ReadStructure(READ_STRUCTURE);
         if (DATA_TYPES.isEmpty()) {
-            DATA_TYPES.addAll(Arrays.asList(IlluminaBasecallsConverter.DATA_TYPES_NO_BARCODE));
+            DATA_TYPES.addAll(Arrays.asList(BasecallsConverter.DATA_TYPES_WITH_BARCODE));
```
Curious why this changed from `NO_BARCODE` to `WITH_BARCODE` and if that was intentional

tfenne commented on src/main/java/picard/illumina/parser/IlluminaDataProviderFactory.java:
```diff
@@ -117,7 +118,7 @@
 
 
-    private List<Integer> availableTiles;
+    private final List<Integer> availableTiles;
```
I wonder everywhere we have List<Integer> for tiles if we can/should just replace to `int[]`?

tfenne commented on src/main/java/picard/illumina/parser/IlluminaDataProviderFactory.java:
```diff
@@ -256,45 +223,75 @@ public void setApplyEamssFiltering(final boolean applyEamssFiltering) {
      * @return An iterator for reading the Illumina basecall output for the lane specified in the constructor.
      */
-    public BaseIlluminaDataProvider makeDataProvider(List<Integer> requestedTiles) {
+    public BaseIlluminaDataProvider makeDataProvider(List<Integer> requestedTiles, File barcodeFile) {
```
This is really weird to me - it takes in a list of tiles and a singular barcode file?  Why is the barcode file even coming from the outside?

jacarey replied:
This is because the new data provider does barcode parsing directly vs deferring to the Parser/Reader. I'll see if I can get around this.

tfenne commented on src/main/java/picard/illumina/parser/IlluminaDataProviderFactory.java:
```diff
@@ -256,45 +223,75 @@ public void setApplyEamssFiltering(final boolean applyEamssFiltering) {
+            final File[] filterFiles = getTiledFiles(laneDir, laneTileRegex);
+
+            IOUtil.assertFilesAreReadable(Arrays.asList(filterFiles));
+            //UGH new provider does not support multiple tiles at a time.
+            return new NewIlluminaDataProvider(cbcls, locs, filterFiles, lane, requestedTiles.get(0), outputMapping, barcodeFile);
```
So I think we need to fix this.  The options I see are:
1. Narrow the API to only allow requesting a single tile at a time and fix anywhere that does anything else
2. Fix NewIlluminaDataProvider to work for multiple tiles
3. Add a `MultiTileIlluminaDataProvider` that can wrap an existing DataProvider/Factory and add multi-tile support.

Thoughts?

jacarey replied:
I went with 2. It seemed like the most straightforward change, but I could be convinced otherwise if you think the other methods have merit. The only place in the code base that actually uses a multiple tile provider is ExtractIlluminaBarcodes, so doing 1. could likely be easier, but this seems the most flexible.

tfenne reviewed (changes requested):
A bunch of comments.  I really do think we should javadoc the hell out of anything we touched so we're leaving things better than we found them.  A few ideas also for where else we might seek to improve performance.

Commit:
Update to BasecallsConverter abstract class.

Commit:
Added UnsortedBasecallsConverter.

tfenne committed:
Speed up encoding of Casava read names.

tfenne committed:
Attempt to make things much more efficient when not sorting by bypassing the conversion from ClusterData into FastqRecordsForCluster.
BclReader now reads from it's streams into a small buffer to speed things up.

tfenne committed:
Used pre-generated lookups for decoding BCL bases and quals.

Commit:
Review changes.
Added ClusterDataCodec so that SortedBasecallsConverter can use ClusterWriter instead of FastqRecordWriter for SortingCollections.

Commit:
Added basic write threading to sorted converter.

Commit:
Added basic write threading to unsorted converter.

tfenne committed:
Fix to ensure that IOUtil.openFileForWriting() will respect CommandLineProgram.COMPRESSION_LEVEL when writing gzipped files.

Commit:
Fix bug in thread count setting.

Commit:
Small performance tweaks.

Commit:
Change tests using deprecated `MAX_READS_IN_RAM_PER_TILE`.
Adjust error checking on executors to avoid potential race conditions.

Commit:
Apply suggestions from code review

Co-authored-by: George Grant <ggrant@broadinstitute.org>
Co-authored-by: Nils Homer <nh13@users.noreply.github.com>
Co-authored-by: kachulis <39926576+kachulis@users.noreply.github.com>

Commit:
Review changes.
Added a lot of unsorted test data for IlluminaBasecallsToSam.

Commit:
Remove unsorted test data.
Fix for potential race condition.

Commit:
Added a delightful space.

Commit:
Merge branch 'master' into jc_refactor

Commit:
Test fixups
